tensorflow_datasets/scripts/replace_fake_images.py

```diff
@@ -115,9 +117,43 @@ def rewrite_tar(root_dir, tar_filepath):
 
     rewrite_dir(temp_dir)  # Recursivelly compress the archive content
 
-    # Converting into tarfile again to decrease the space taken by the file-
+    os.chdir(temp_dir) # Change to temporary directory so that temp directory name does not affect original filenames
     with tarfile.open(tar_filepath, 'w' + extension) as tar:
-      tar.add(temp_dir, recursive=True)
+      tar_add(tar, '.', recursive=True) # Call custom recursive tar add operation
+
+
+def tar_add(tar_file, name, arcname=None, recursive=True):
```

Comment:
I am confused by this function.
* Why use `bltn_open` instead of `open` ?
* Why have `recursive` arg if this is always True ?
* The `os.chdir(` seems like a hack and should not be used. Instead, why not simply truncate the name ?
* Why checking `isfile()` inside the `isdir()` scope ?

I don't understand why you need this complicated function.
Can't you just use something like ?

```python
with tarfile.open(tar_filepath, 'w' + extension) as tar:
  tar.tar_file(tmp_dir, arcname='', recursive=True)
```
Or if the above function do not work:
```python
with tarfile.open(tar_filepath, 'w' + extension) as tar:
  with filename in os.path.listdir(tmp_dir):
    tar.tar_file(os.path.join(tmp_dir, filename), arcname=filename, recursive=True)
```

Reply:
Hey you are absolutely right, there is no need for this complication. I was under the wrong impression that tarfile.getnames() should only return filenames and not directories.(Consequence of overthinking). 
The code is complicated because it is from [https://github.com/python/cpython/blob/master/Lib/tarfile.py](url). (The python library tarfile.py)
You are right about both, os.chdir is hacky and recursive=True is not required.
My sincerest apologies for wasting your time.

Your 2nd suggestion (which I had obviously tried and rejected foolishly as the tarfile still included directories)
```python3
    with tarfile.open(tar_filepath, 'w' + extension) as tar:
      for filename in os.listdir(temp_dir):
        tar.add(os.path.join(temp_dir, filename), arcname=filename,
                recursive=True)
```
works perfectly fine. I have run tests on random datasets from images, audio and text. All test pass. I have made this change along with the new fake_data_dir output in the new commit.
Also no extra code needs to be written for '.tgz' files as the tarfile library takes care of the extension. ( i have tested the latest commit on this as well)

Reply:
Thank you for the update. Why haven't you tried the first suggestion ?
```
with tarfile.open(tar_filepath, 'w' + extension) as tar:
  tar.tar_file(tmp_dir, arcname='', recursive=True)
```

This seems simpler, no ? Was there an issue with this one ?

Reply:
My apologies, I don't know why I am embarrassing myself !
Yes the first solution is way simpler and easier to read,  also solves the problem of "./" being added to each component file.
I have tried your first suggestion and all random tests pass. Also tested '.tgz' extension check which passes as well with the latest commit.

Again I apologise for wasting precious time. I think I should calm down with the commits.

Reply:
Also a question, do you think using shutil.make_archive for zip files and low-level API for tar files is OK ? 
Something similar with zipfile library can also be done, however unlike tarfile there is no recursive write option. Should I try to eliminate shutil for zipfiles ?

Reply:
So for example this piece of code (modified slightly from the previous version):
```python3
    if zip_filepath.lower().endswith('zip'): # new line
      with zipfile.ZipFile(zip_filepath, 'w', compression=zipfile.ZIP_DEFLATED) as zip_file:
        for file_dir, _, files in os.walk(temp_dir):
          for file in files:
            file_path = os.path.join(file_dir, file)
            arcname=os.path.relpath(file_path, temp_dir) # new line
            zip_file.write(file_path, arcname=arcname)
```
does the same job as shutil.make_archive(). I have tested for this version as well and all tests for datasets with zip file PASS. Should I switch this back as well ?